server/src/main/java/io/crate/execution/dml/RawIndexer.java

```diff
@@ -115,11 +114,12 @@ public List<Reference> collectSchemaUpdates(IndexItem item) throws IOException {
         assert numExtra == nonDeterministicSynthetics.size() : "Insert columns/values expansion must be done in sync";
 
         Object[] insertValues = new Object[doc.size() + numExtra];
-        Iterator<Object> iterator = doc.values().iterator();
         List<Reference> columns = currentRowIndexer.columns();
         for (int i = 0; i < doc.size(); i++) {
             Reference reference = columns.get(i);
-            Object value = iterator.next();
+            // JSON file can have rows with same keys but in different order.
+            // We rely on reused currentRowIndexer's target columns order when fetching values from every row.
+            Object value = doc.get(reference.column().name()); // Not using FQN, we are fetching only top-level columns.
```

Comment:
Can't this lead to issues when the related column doesn't exists? E.g. when a `DEFAULT` clause is defined on that column, it would write a `NULL` value and such not using the default expression.

Reply:
No, such columns are handled by Indexer.synthetics (generated if not provided by the currentRowIndexer). If provided they will be validated currentRowIndexer but "provided" is basically covered here.

Another exception when synthetics are part of target columns is non-deterministic expressions but they are generated on primary and copied to replica. it's unrelated to JSON file values.
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/RawIndexer.java#L131

Reply:
Can `Map<String, Object> doc = XContentHelper.convertToMap(JsonXContent.JSON_XCONTENT, raw, true);` further above now get changed to set ordered=false ?

Reply:
changed to `ordered=false`

Reply:
Had to revert it, otherwise test_copy_preserves_implied_top_level_column_order is failing.For json we use keys order in the first row as source of truth for positions of dynamically created columns.